Reworked the application shell

spoenemann opened the pull request:
First step towards #1025.

I added a bottom panel to the application shell. The "Open Problems View" command now opens the view in that panel.

I also refactored and cleaned up some of the code in `shell.ts` and `sidepanel.css`.

akosyakov commented:
Nice! Some things to take care:
- [ ] the bottom bar should look & feel like a tab bar in the main area, i.e. ability to close individual widgets, the context menu, black background
- [ ] there is empty space in a tab caption of sidebars, apparently for icons
- [ ] would be nice to open terminals in the bottom bar

spoenemann commented:
* [ ] Add open commands for Files, Git, Extensions and Outline to the View menu.

svenefftinge commented:
This will fix https://github.com/theia-ide/theia/issues/533

spoenemann commented on examples/browser/test/bottom-panel/bottom-panel.ts:
```diff
@@ -0,0 +1,47 @@
+import "webdriverio";
```
&#9989;

svenefftinge commented on packages/core/src/browser/shell.ts:
```diff
@@ -30,104 +32,109 @@ import { Saveable } from "./saveable";
-    activeWidgets?: Widget[]
-}
+export const MAIN_AREA_TABBAR_CONTEXT_MENU: MenuPath = ['main-area-tabbar-context-menu'];
+export const SIDE_AREA_TABBAR_CONTEXT_MENU: MenuPath = ['side-area-tabbar-context-menu'];
```
we should have one context menu for the tabs and remove certain commands by implementing `isVisible`

svenefftinge commented on packages/navigator/src/browser/navigator-frontend-module.ts:
```diff
@@ -13,15 +13,23 @@ import { NavigatorMenuContribution } from './navigator-menu';
+    bind(KeybindingContribution).toDynamicValue(c => c.container.get(FileNavigatorContribution));
     bind(MenuContribution).to(NavigatorMenuContribution).inSingletonScope();
+
     bind(FileNavigatorWidget).toDynamicValue(ctx =>
```
could you remove this binding, so it is not exposed and accidentally used?

spoenemann replied:
There are bindings for other widgets, too: `problem-frontend-module.ts`, `git-frontend-module.ts`, `extension-frontend-module.ts`. What about those?

akosyakov commented on packages/navigator/src/browser/navigator-container.ts:
```diff
@@ -41,5 +45,25 @@ export function createFileNavigatorContainer(parent: interfaces.Container): Cont
-    return createFileNavigatorContainer(parent).get(FileNavigatorWidget);
+    const fileNavigator = createFileNavigatorContainer(parent).get(FileNavigatorWidget);
+
+    const selectionService = parent.get(SelectionService);
```
Could we move this logic to `FileNavigatorWidget` constructor?

svenefftinge commented:
When I first open it (or use reset layout) it shows an empty expanded bottom panel:

<img width="974" alt="screen shot 2018-01-10 at 16 24 29" src="https://user-images.githubusercontent.com/372735/34780246-e3995cd8-f622-11e7-8c1c-4ca8ac82b76d.png">

It should be collapsed and when empty not visible at all.

svenefftinge commented:
The commands should be renamed to `Toggle <View>` and they should close the view, if the view is currently focussed, otherwise open and focus.

Terminals are not singleton widgets and therefore don't really make sense in the 'view' space. I'd suggest to move the command back to `File` and call it `New Terminal`.

For the layout we should just change it so that the side panels get the full height and the bottom and main area the rest.

akosyakov commented on packages/extension-manager/src/browser/extension-contribution.ts:
```diff
@@ -34,11 +49,4 @@ export class ExtensionContribution implements FrontendApplicationContribution {
-    async initializeLayout(app: FrontendApplication): Promise<void> {
-        const extensionWidget = await this.widgetManager.getOrCreateWidget('extensions');
-        app.shell.addToLeftArea(extensionWidget, {
-            rank: 300
```
What about rank how the default order could be defined? the same for Git

akosyakov commented on packages/navigator/src/browser/navigator-widget.ts:
```diff
@@ -30,12 +33,33 @@ export class FileNavigatorWidget extends FileTreeWidget {
+        this.initialize(selectionService, workspaceService, labelProvider);
+    }
+
+    protected initialize(selectionService: SelectionService, workspaceService: WorkspaceService, labelProvider: LabelProvider): void {
```
args can be properties instead?

akosyakov commented:
I got the following error trying it out:
```
Uncaught (in promise) TypeError: Cannot set property 'parent' of undefined
    at SideBarHandler.addWidget (shell.ts:988)
    at SideBarHandler.setLayoutData (shell.ts:937)
    at ApplicationShell.setLayoutData (shell.ts:313)
    at shell-layout-restorer.ts:71
    at <anonymous>
```

akosyakov commented:
Probably out of the scope: keybindings are displayed only in the selection menu, should not they be displayed in all menus? e.g in `View` it will be handy

spoenemann commented:
> Probably out of the scope: keybindings are displayed only in the selection menu, should not they be displayed in all menus?

#490

svenefftinge commented:
Looks like the test failures are 'real'

akosyakov commented on packages/core/src/browser/shell/view-contribution.ts:
```diff
@@ -0,0 +1,111 @@
+
+    async openView(args: Partial<OpenViewArguments> = {}): Promise<T> {
+        const widget = await this.widget;
+        const tabBar = this.shell.gettTabBarFor(widget);
```
typo: `gettTabBarFor` ?

spoenemann commented:
In the course of reacting to review comments, I have refactored and cleaned up a lot of code, so this PR has now become much more than just adding a bottom panel. It's a complete rework of the application shell and its API.

Note that I have touched almost all public methods of `ApplicationShell`, so this PR is a breaking change.

hexa00 commented on packages/terminal/src/browser/terminal-frontend-contribution.ts:
```diff
@@ -84,10 +84,9 @@ export class TerminalFrontendContribution implements CommandContribution, MenuCo
             keybindings.registerKeybindings({
                 commandId: KeybindingRegistry.PASSTHROUGH_PSEUDO_COMMAND,
-                keyCode: KeyCode.createKeyCode({ first: k, modifiers: [modifier] }),
+                keyCode: KeyCode.createKeyCode({ first: k, modifiers: [Modifier.CTRL] }),
```
why is that changed?

spoenemann replied:
I extracted the check `isOSX ? Modifier.M4 : Modifier.M1` to a new constant that can (should) be used where CTRL is needed independently of the platform. The behavior is unchanged.

hexa00 replied:
OK indeed nice :)

hexa00 commented:
Looks great in general! 

It may need quite a lot of manual testing with all it changes, I'll try to play around with it today.

spoenemann commented:
There are new key bindings:

* M1 + Shift+ E: toggle Files
* Ctrl + Shift + G: toggle Git
* M1 + Shift + X: toggle Extensions
* Ctrl + Shift + O: toggle Outline
* M1 + Shift + M: toggle Problems
* Alt + C: collapse side panel
* Alt + Shift + C: collapse all side panels

hexa00 commented:
Feel free to add that to the commit log :)

AlexTugarev commented:
@spoenemann just FYI, the branch needs an update. `@theia/output` was added and needs to learn new API.

AlexTugarev commented:
@spoenemann, I'd like to propose adding optional options to main panel, similar to what I've tried out for markdown preview extension, cf. https://github.com/theia-ide/theia/pull/1051/commits/0cdaa1d5c90c79adb73672d093b3c9c805c081cc and demo: https://github.com/theia-ide/theia/pull/1051#issue-288519297

akosyakov commented on packages/core/src/browser/shell/view-contribution.ts:
```diff
@@ -0,0 +1,111 @@
+                ...args
+            };
+            this.shell.addWidget(widget, widgetArgs);
+        } else if (tabBar.currentTitle === widget.title && args.toggle) {
```
`widget` has `isVisible` attribute, could it be used here?

akosyakov replied:
or `this.shell.currentWidget === widget` it should mean that a widget is attached, visible and in focus

spoenemann replied:
The widget should be closed when it is already visible in order to realize the toggle behavior. The tab bars indicate which widgets are visible.

spoenemann commented:
> `@theia/output` was added and needs to learn new API.

&#9989;

> I'd like to propose adding optional options to main panel, similar to what I've tried out for markdown preview extension

&#9989;

Also fixed an issue regarding shell layout storage where the order of widgets was not consistent after reloading.

svenefftinge commented:
Nice!, please merge.
